### DISCUSSION TITLE
Unify Device and VirtualMachine in a single table. Differentiate between the 2 using `is_virtual` flag.

### DISCUSSION THREAD

**Discussion by slappey-ibkr** [nautobot/nautobot#1015] [Ideas]
### Environment
* Python version: 3.6
* Nautobot version: 1.1.4

### Proposed Functionality
Add `is_virtual` flag to Device model and deprecate VirtualMachine or make it a proxy model back to Device with `is_virtual` hardcoded to True.  I know there's more to it than that but I'd love to hear everyone's thoughts on this.

### Use Case
The problem with having 2 tables is that we must know if a host is physical or virtual when running remote jobs.  Otherwise, we have to query both tables until we find the host we’re looking for.  We must also know if a host is a physical or virtual when creating relations to objects such as IPAddress.  The proposed functionality would unify all devices in a single table.

### Database Changes
1. Add `is_virtual` flag to Device
2. (not required) Update VirtualMachine model to be a proxy to Device with `is_virtual` hardcoded to True.
3. (not required) Create new model PhysicalMachine as a proxy to Device with `is_virtual` hardcoded to False.

**if device.is_virtual == False:**
- device is a physical machine
- validation of device.rack is enabled and rack is visible in UI
- validation of device.position is enabled and rack position is visible in UI
- device.cluster becomes the cluster this physical machine is a member of
- device.interfaces.type choices are enabled and are visible in UI

**if device.is_virtual == True:**
- device is a virtual machine
- validation of device.rack is disabled and is hidden in UI
- validation of device.position is disabled and is hidden in UI
- device.cluster becomes the cluster this virtual machine is running on
- device.interfaces.type choices are disabled, value defaults to virtual and is hidden in UI

**Example Queries**
device_count = Device.objects.filter(is_virtual=False).count()
vm_count = Device.objects.filter(is_virtual=True).count()

cluster = Cluster.objects.first()
cluster_devices = cluster.devices.filter(is_virtual=False)
cluster_vms = cluster.devices.filter(is_virtual=True)

### External Dependencies
None

- - - -

*Again, I know there's more to it than this but I'd love to kick off the conversation and hear everyone's thoughts.*

**Comment by glennmatthews** [comment]
Thank you! As you suggested, this is probably an idea that makes sense to have a great deal more conversation about before making any design decisions. As such I'm going to move this issue to the Discussions for now.

**Comment by glennmatthews** [comment]
We're planning to move ahead with this idea in an upcoming release. As such we've moved this back to an open issue -- please refer to #1178, and please provide any comments or feedback there.
